docs/porting/target/static_pinmap.md

```diff
@@ -0,0 +1,183 @@
+<h2 id="static-pinmap-port">Static Pinmap extension</h2>
+
+The **Static Pinmap extension** allows the peripheral configuration (pin/periheral/function) to be staticly specified in the HAL API function.
+
+### Overview and background
+
+In modern MCUs peripherals often can be mapped to different pins and each pin can have multiple functions. Mbed supports dynamic pin mapping, meaning that pins can be reconfigured at run time to be used by different driver. That provides great flexibility, but it's not free. There's non trivial ROM cost to maintain the pinmap tables and infrastructure to parse it. In some use cases this flexibility is worth the cost. Quite often pin configuration is frozen at hw design stage and doesn't require runtime modification. Shifting this configuration to compile time will allow us free memory associated with the dynamic approach.
+
+HAL APIs making use of pins take these pins in their constructor and use those pins to lookup which peripheral/function to use. The process of looking up the peripheral/function requires there to be a pinmap table that maps pins to peripherals/functions. This pinmap table takes up ROM which could be saved if the pinmap wasn't used. Static pinmap extension provides additional HAL API/constructors which takes pinmap as a parameter where pin/peripheral/function is specified staticly and there is no need to use the pinmap tables.
+
+Supported peripherals:
+ - `PWM`
```

Comment:
What about digitalio? I mean it would make a lot of sense to say `Pins p23, p24 DigitalOut; p12 DigitalIn; Pins p34, 35 UART; all the other pins not used` In this case we would only need descriptors for 5pins and ignore the remaining ones saving further memory.

Reply:
I'm not sure which, if any, platforms would benefit.

Most perform a fairly simple computation from pin number to GPIO bit/peripheral, and the static information they write into their `gpio_t` is a bit more varied than doing the basic muxing of "set function 3" when selecting an alternate function - they're not using our `PinMap` type. The "pin/peripheral/function" 3-ple probably wouldn't cut it.

What might make sense is the option of a target `constexpr` `gpio_t`-returning function, and then have something like

    struct gpio_pinmap_t {
    #ifdef HAVE_STATIC_GPIO
        gpio_t gpio;
    #else
        PinName pin;
    #endif
    };
    
    constexpr gpio_pinmap_t gpio_get_pinmap(PinName pin)
    {
    #ifdef HAVE_STATIC_GPIO
        return { hal_get_gpio(PinName pin) };
    #else
        return { pin };
    #endif
    }

    DigitalIn::DigitalIn(const gpio_pinmap_t &map)
    #ifdef HAVE_STATIC_GPIO
        : gpio(map.gpio)
    #endif
    {
    #ifndef HAVE_STATIC_GPIO
        gpio_init(&gpio, map.pin);
    #endif
        ...
    }


    constexpr gpio_pinmap_t pin_a_map = gpio_get_pinmap(PIN_A);
    DigitalIn pin_a(pin_a_map);

No idea whether that really saves anything significant in practice. It might be worse on some platforms.

Reply:
As far as I remember only some Freescale targets have pin-maps for GPIOs, but even these targets do not use them during initialization. Probably they are redundant.